Add license key support to MaxMind GeoLite2 database downloads

ObliviousHarmony opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Due to the CCPA, MaxMind has removed public access to their GeoLite databases. In order to download the database you must now create an account and generate a license key. A side-effect of this change however is that the GeoLite2 database download in WooCommerce is non-functional. This PR adds support for a license key, as well as lays the groundwork for the MaxMind integration to be moved into a feature plugin.

Closes #25361 and #25370.

### How to test the changes in this Pull Request:

1. Make sure there is no "GeoLite2-Country.mmdb" file in your uploads directory.
2. In your store settings, change the "Default customer location" to one of the Geolocate options. (If you already have been set to one, change to something else and then to a geolocate option.)
3. Without the patch, note that there is no database in your uploads directory. If you check your logs, there will be a geolocation error indicating that it was unable to download the database.
4. Check out the branch and visit the Settings > Integration section. After creating an account on MaxMind's website and generating a key, place that key into the setting and save.
5. Change the default customer location again.
6. Note the archive in the uploads folder.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

ObliviousHarmony committed:
Moved the MaxMind DB Reader library into composer and pinned the latest version

ObliviousHarmony committed:
Added support for creating default integrations

ObliviousHarmony committed:
Added a default integration settings page for managing MaxMind Geolocation services

ObliviousHarmony committed:
Added support for downloading and extracting the GeoLite2 database

ObliviousHarmony committed:
Added filterable support for fetching the database path

This replaces the corresponding geolocation class functionality. We are deprecating the woocommerce_geolocation_local_database_path filter in favor of the woocommerce_maxmind_geolocation_databse_path filter. Since we want the option to turn this into a feature plugin, it's wise to make sure nothing is too generic.

ObliviousHarmony committed:
Made the database service class methods static for convenience \

ObliviousHarmony committed:
Fixed a permissions conflict caused by the dummy GeoLite2 archive's date

The test would fail if the web server had stored the data already!

ObliviousHarmony committed:
Fixed the validation for MaxMind license keys

ObliviousHarmony committed:
Replaced the Geolocation database fetching functionality with MaxMind integration

ObliviousHarmony committed:
Added back the WC_Geolocation's get_local_database_path method since it was publicly used

claudiosanches commented on includes/class-wc-integrations.php:
```diff
@@ -48,4 +48,18 @@ public function __construct() {
+	private function get_default_integrations() {
+		$default_integrations = array();
+
+		include_once WC_ABSPATH . 'includes/integrations/maxmind-geolocation/class-wc-maxmind-geolocation-integration.php';
```
It should be part of the `WC_Autoloader` and load any integration class inside `integrations` directory.

claudiosanches replied:
Doing it we can just list the default integrations array inside `__construct` like we does on `WC_Payment_Gateways` class.

ObliviousHarmony replied:
Agreed, thanks!

claudiosanches commented on includes/integrations/maxmind-geolocation/class-wc-maxmind-geolocation-database.php:
```diff
@@ -0,0 +1,115 @@
+	 * @return string|WP_Error The path to the database file or an error if invalid.
+	 */
+	public static function download_database( $license_key ) {
+		$download_uri  = 'https://download.maxmind.com/app/geoip_download?';
```
You can replace this code for:

```php
$download_uri = add_query_arg( 
	array(
		'edition_id'  => self::DATABASE,
		'license_key' => $license_key,
		'suffix'      => 'tar.gz',
	),
	'https://download.maxmind.com/app/geoip_download'
 );
```

ObliviousHarmony replied:
That's a useful helper, thanks!

claudiosanches commented on tests/unit-tests/integrations/maxmind-geolocation/class-wc-tests-maxmind-database.php:
```diff
@@ -0,0 +1,135 @@
+	 * Tests that the database download works as expected.
+	 */
+	public function test_download_database_works() {
+		$expected_database = '/tmp/GeoLite2-Country_20200100/GeoLite2-Country.mmdb';
```
This directory name may change with future releases of GeoLite and will make this test to fail.

ObliviousHarmony replied:
This test doesn't actually perform the download, but rather "downloads" a file from the `tests/data` directory. Since we can't distribute real MaxMind archives in the repository, this is a mock of the downloaded archive, with fixed file paths.

claudiosanches replied:
Cool!

claudiosanches commented on includes/class-wc-geolocation.php:
```diff
@@ -253,44 +253,32 @@ public static function update_database() {
-				// Make sure target dir exists.
-				$wp_filesystem->mkdir( dirname( $target_database_path ) );
+		// There's no need to update the database with no geolocation configured.
+		$license_key = ( new WC_MaxMind_Geolocation_Integration() )->get_option( 'license_key' );
```
This keep the dependency on this class, let's change it into a hook, so in the integration we can hook into it and download the file.

claudiosanches commented on includes/class-wc-geolocation.php:
```diff
@@ -207,7 +207,7 @@ public static function geolocate_ip( $ip_address = '', $fallback = false, $api_f
 			} else {
 				$ip_address = $ip_address ? $ip_address : self::get_ip_address();
-				$database   = self::get_local_database_path();
+				$database   = WC_MaxMind_Geolocation_Database::get_database_path();
```
This also could be changed by a filter, so we can filter by the integration.
Also we can remove `supports_geolite2` and `geolocate_via_db`.

claudiosanches replied:
The idea is to make this class let's directly dependent on an integration.

claudiosanches replied:
Also missing put a deprecate notice about `GEOLITE2_DB`.

ObliviousHarmony replied:
Good catch!

claudiosanches commented on includes/class-wc-geolite-integration.php:
```diff
@@ -38,10 +38,6 @@ class WC_Geolite_Integration {
 	public function __construct( $database ) {
 		$this->database = $database;
-
-		if ( ! class_exists( 'MaxMind\\Db\\Reader', false ) ) {
```
We could move/copy `get_country_iso()` method to `WC_MaxMind_Geolocation_Database` or into the integration class, and deprecate this class.

ObliviousHarmony replied:
Agreed!

claudiosanches commented on includes/integrations/maxmind-geolocation/class-wc-maxmind-geolocation-integration.php:
```diff
@@ -0,0 +1,80 @@
+			'license_key' => array(
+				'title'       => __( 'MaxMind License Key', 'woocommerce' ),
+				'type'        => 'password',
+				'description' => __( 'The key that will be used when dealing with MaxMind Geolocation services. You can generate a key in your account after creating an account on MaxMind\'s website.', 'woocommerce' ),
```
We should include a link to the registration or probably better for a documentation of how to do it.

ObliviousHarmony replied:
How would you feel about linking to [MaxMind's support documentation page on the subject?](https://support.maxmind.com/account-faq/account-related/how-do-i-generate-a-license-key/)

claudiosanches replied:
Yep, sounds fine.

claudiosanches reviewed (dismissed):
Great work!

There's a few things to change or improve, also it's missing some notice for the user in case Geolocation is enabled and there's no license key integration settings, this notice could be "dismissible" or not, but at least should shows up once for shop owns set up it.

ObliviousHarmony committed:
Moved the loading of integration classes into the WC_Autoloader

ObliviousHarmony committed:
Fixed the erroneous return in WC_Geolocation::get_local_database_path

ObliviousHarmony committed:
Simplified the URL generation for geolocation database downloads

ObliviousHarmony committed:
Added a deprecation notice for the GEOLITE2_DB constant

claudiosanches commented on includes/class-wc-geolocation.php:
```diff
@@ -237,7 +238,8 @@ public static function geolocate_ip( $ip_address = '', $fallback = false, $api_f
 	 */
 	public static function get_local_database_path( $deprecated = '2' ) {
-		return apply_filters( 'woocommerce_geolocation_local_database_path', WP_CONTENT_DIR . '/uploads/GeoLite2-Country.mmdb', $deprecated );
+		require_once WC_ABSPATH . 'includes/integrations/maxmind-geolocation/class-wc-integration-maxmind-geolocation-database.php';
```
This doesn't sound correct, since `WC_Autoloader` should load this integration before `includes/class-wc-geolocation.php` is included.

claudiosanches replied:
Seems like `class-wc-integrations.php` is included after `class-wc-geolocation.php`, this doesn't sound correct, integrations suppose to work everywhere.

claudiosanches commented on includes/integrations/maxmind-geolocation/class-wc-integration-maxmind-geolocation-database.php:
```diff
@@ -0,0 +1,115 @@
+		$download_uri = add_query_arg(
+			array(
+				'edition_id'  => self::DATABASE,
+				'license_key' => $license_key,
```
I think we could sanitize this `license_key` with `wc_clean()`, or the escape the URL with `esc_url_raw()`.

ObliviousHarmony replied:
Ah, I didn’t realize that the query arguments wouldn’t be escaped this way. Arguably, I don’t actually think in this case that there are any security problems though, and the license keys are ASCII it appears. Can’t hurt to add though, of course.

ObliviousHarmony committed:
Added sanitization to the license key in download URLs

ObliviousHarmony committed:
Removed the unnecessary GeoLite2 version check

ObliviousHarmony committed:
Moved the loading of the database class into the integration

ObliviousHarmony committed:
Deprecated the WC_Geolocation::get_local_database_path method

ObliviousHarmony committed:
Added a documentation link to the license key setting description

ObliviousHarmony committed:
Moved the database update from the WC_Geolocation class to the MaxMind-specific integration

ObliviousHarmony committed:
Moved the MaxMind geolocation lookup into the integration class

ObliviousHarmony committed:
Fixed the response for the geolocation filter

ObliviousHarmony committed:
Refactored the IP geolocation method

stodorovic commented on includes/class-wc-geolocation.php:
```diff
@@ -194,6 +192,27 @@ public static function update_database() {
+	private static function get_country_code_from_headers() {
+		$country_code = false;
+
+		if ( ! empty( $_SERVER['HTTP_CF_IPCOUNTRY'] ) ) { // WPCS: input var ok, CSRF ok.
```
I think that will be nice to you add _MM_COUNTRY_CODE_ which is often default environment variable for [mod_maxminddb](https://github.com/maxmind/mod_maxminddb).

Maybe it's possible to use _foreach_ instead of few _if conditionals_...  (I don't know is it fine by your coding standards). Something like this:

```php
$var_names = ( 'HTTP_CF_IPCOUNTRY', 'MM_COUNTRY_CODE', 'GEOIP_COUNTRY_CODE', 'HTTP_X_COUNTRY_CODE' );

foreach ( $var_names as $var ) {
	if ( ! empty( $_SERVER[ $var ] ) ) {
		....
		break;
	}
}
```

ObliviousHarmony replied:
Thank you for this suggestion, I've gone ahead and made a change.

ObliviousHarmony committed:
Added support for mod_maxminddb's MM_COUNTRY_CODE server variable

ObliviousHarmony committed:
Refactored the MaxMind database class to not use static methods in order to make testing the integration class easier

ObliviousHarmony committed:
Fixed a bug where empty database paths might break the update

ObliviousHarmony committed:
Added unit tests for the MaxMind integration class

claudiosanches commented on includes/class-wc-geolocation.php:
```diff
@@ -81,53 +72,6 @@ private static function is_geolocation_enabled( $current_settings ) {
-	 * @param string $default_customer_address current value.
-	 * @return string
-	 */
-	public static function disable_geolocation_on_legacy_php( $default_customer_address ) {
```
We can't remove those public methods for backwards compatibility reasons.

claudiosanches committed:
Fixed missing methods, added deprecated notices and @since

claudiosanches committed:
Prevent fatal error if MaxMind Reader is not installed

claudiosanches committed:
Fixed fatal error in system status page

claudiosanches committed:
Removed legacy code

peterfabian reviewed (commented):
Great job, thanks! I tested it in different scenarios and it seems to work fine. 
Added just two minor notes, will look at the code a bit more again.

claudiosanches committed:
Updated integration ID and included methods to display notices

claudiosanches committed:
Fixed integration ID

claudiosanches committed:
Added missing license key notice and triggers

claudiosanches commented on includes/integrations/maxmind-geolocation/class-wc-integration-maxmind-geolocation.php:
```diff
@@ -0,0 +1,168 @@
+						'woocommerce'
+					),
+					'https://support.maxmind.com/account-faq/account-related/how-do-i-generate-a-license-key/'
+				),
```
I think we could link into a documentation on docs.woocommerce.com.
So we can give more details of how to generate.

peterfabian replied:
Yeah, that might be even better.

claudiosanches replied:
I'll change it for a documentation later, before we release. I still need a link for it.

claudiosanches commented on tests/unit-tests/geolocation/class-wc-tests-geolite-integration.php:
```diff
@@ -1,40 +0,0 @@
-		// Check for non-valid IP.
-		$this->assertEquals( '', $geolite->get_country_iso( 'foobar' ) );
-	}
-}
```
Since it's deprecated we should not update that class anymore, also there's no reason for use this class anymore.
But I agree that we could keep this unit tests in place while we don't remove that class from core.

claudiosanches replied:
Ohh, I just checked and the current code makes hard to keep this unit test working, currently this test is failing, since it's not possible to download the database without a license key, also that method uses MaxMind Reader library to read the file, we currently have another tests for it too, and if we move that code to read a custom database in our unit tests will not make any sense to keep this test, since is more like testing the MaxMind Reader Library than code on WooCommerce.

ObliviousHarmony replied:
This was my thinking in removing the test.

peterfabian replied:
Since the geo database is now included with the tests in `tests/data`, does it matter that the license code is not there? I thought the class should still continue to work even if it's deprecated, no?

I agree there is not that much of our code to test, but the code for the method `get_country_iso` hasn't changed, so why it made sense to test it before, but it doesn't make sense to test it now? I still think it's good to have the test to make sure we don't break bw compat of this class until we decide to delete it.

ObliviousHarmony replied:
The included database isn’t actually real. We can’t include the GeoLite database with the repository, it’s an empty file for testing the service.

ObliviousHarmony replied:
The included database isn’t actually real. We can’t include the GeoLite database with the repository, it’s an empty file for testing the service.

peterfabian replied:
ah, I see. Haven't realized that, apologies. Let's remove it, then.

claudiosanches committed:
Remove alternatives for deprecated methods

ObliviousHarmony committed:
Adjusted the MaxMind license key notice to be clearer

oscargare commented on includes/integrations/maxmind-geolocation/class-wc-integration-maxmind-geolocation-database.php:
```diff
@@ -0,0 +1,115 @@
+	 * @return string The local database path.
+	 */
+	public static function get_database_path() {
+		$database_path = WP_CONTENT_DIR . '/uploads/' . self::DATABASE . self::DATABASE_EXTENSION;
```
Hi, I think the database should be stored in a non-public directory to comply with the [Site License](https://www.maxmind.com/en/site-license-overview).

ObliviousHarmony committed:
Added support for a randomized MaxMind database file prefix

ObliviousHarmony committed:
Removed permanently failing unit test for deprecated GeoLite2 library integration

ObliviousHarmony committed:
Moved the database file into the woocommerce_uploads folder

ObliviousHarmony committed:
Added support for non-standard upload folder paths

ObliviousHarmony committed:
Added upgrade step for removing the legacy MaxMind database

claudiosanches committed:
Remove notice when license key is present

ObliviousHarmony committed:
Added a database option to allow for selecting different MaxMind databases

ObliviousHarmony committed:
Revert "Added a database option to allow for selecting different MaxMind databases"

This reverts commit d32c470cea3fe7b89f2e0610b95573f2453a0d25.

ObliviousHarmony committed:
Added a note that the MaxMind integration will only do country lookups.

claudiosanches committed:
Removed MaxMind from System Status page

ObliviousHarmony committed:
Changed the geoip update cron to run every 15 days to better comply with MaxMind's TOS

ObliviousHarmony committed:
Added the database path to the integration page for documentation purposes

claudiosanches committed:
Moved HTML to a single file

vedanshujain commented on includes/admin/class-wc-admin-notices.php:
```diff
@@ -87,6 +88,7 @@ public static function reset_admin_notices() {
 		self::add_wc_admin_feature_plugin_notice();
 		self::add_notice( 'template_files' );
 		self::add_min_version_notice();
+		self::add_maxmind_missing_license_key_notice();
```
what do you think about hooking this to `update_option_woocommerce_default_customer_address` in `init` method of this class as well. That way all our notice code will be in one place and we won't need to repeat this in later in `display_missing_license_key_notice` ?

claudiosanches replied:
We run this method on `woocommerce_installed`, so it get triggered after the installation or any update. We plan to move this integration to a feature plugin in the future, so we plan to remove from this class, that's why we are not including all in one place right now.

vedanshujain commented on includes/admin/class-wc-admin-notices.php:
```diff
@@ -428,6 +430,38 @@ public static function wp_php_min_requirements_notice() {
+	 * @since 3.9.0
+	 */
+	public static function maxmind_missing_license_key_notice() {
+		if ( get_user_meta( get_current_user_id(), 'dismissed_maxmind_license_key_notice', true ) ) {
```
also add clause for `woocommerce_maxmind_geolocation_display_notices` filter.

claudiosanches replied:
Good catch!

vedanshujain reviewed (approved):
Awesome work. Left some minor comments, but I am approving because all comments are cosmetic changes and it's working as expected.

claudiosanches commented on includes/admin/settings/class-wc-settings-general.php:
```diff
@@ -185,7 +174,12 @@ public function get_settings() {
 					'type'     => 'select',
 					'class'    => 'wc-enhanced-select',
-					'options'  => $woocommerce_default_customer_address_options,
+					'options'  => array(
```
I agree, `geolocation` shouldn't be the best option for everyone by default.
@ObliviousHarmony do you agree?

ObliviousHarmony replied:
I agree.

claudiosanches commented on includes/class-wc-geolocation.php:
```diff
@@ -368,6 +280,50 @@ private static function geolocate_via_api( $ip_address ) {
+	 * @deprecated 3.9.0
+	 * @return null
+	 */
+	public static function init() {
```
It's deprecated, so it's on the bottom, we should remove it in future versions.

claudiosanches committed:
Apply woocommerce_maxmind_geolocation_display_notices filter in notices class

ObliviousHarmony commented on includes/class-wc-geolocation.php:
```diff
@@ -192,122 +136,90 @@ public static function get_external_ip_address() {
-					$country_code = '';
-				}
+			// Filter to allow custom geolocation of the IP address.
+			$country_code = apply_filters( 'woocommerce_geolocate_ip', $country_code, $ip_address, $fallback, $api_fallback );
```
From a performance standpoint, it’s ideal that we do these inexpensive lookups before deferring to the potentially expensive hook. In our case, the integration uses the hook to perform its lookup. If we are worried about the comparability of it, we could leave this hook at the top instead and add another, but the naming of the hook might be confusing.

claudiosanches replied:
@ObliviousHarmony I agree with @vedanshujain, we could keep this filter like before and then introduce a new one at the end of this function, so there's no problem to run `get_country_code_from_headers()` before.

stodorovic replied:
I've tested it and it works on same way as before. I agree that doesn't make to load couple MBs to get country code when I've installed [mod_maxminddb](https://github.com/maxmind/mod_maxminddb) which already set proper environment variable.

claudiosanches replied:
I solved it restoring the old `woocommerce_geolocate_ip` filter and introducing a new filter. `woocommerce_geolocate_ip` will acts as a pre filter if someone is using it, the same as before, but we check for headers before trying to do any request to a database.

claudiosanches committed:
Introduced woocommerce_get_geolocation filter

claudiosanches committed:
Updated implementation of geolite2

claudiosanches committed:
Use shop base address as default customer location

ObliviousHarmony commented on includes/class-wc-geolocation.php:
```diff
@@ -136,25 +136,50 @@ public static function get_external_ip_address() {
+		 * @param string $ip_address  IP Address.
+		 */
+		$geolocation  = apply_filters(
+			'woocommerce_get_geolocation',
```
What about an action of `woocommerce_do_geolocation` or something, given that this is actually doing the geolocation work?

ObliviousHarmony reviewed (commented):
The filtering changes look great, just maybe a different name?

claudiosanches committed:
Fixed conditional code

rodrigoprimo committed:
Fix typo in php docblock

rodrigoprimo committed:
Merge branch 'fix/25361' of github.com:woocommerce/woocommerce into fix/25361

claudiosanches committed:
Introduced woocommerce_maxmind_geolocation_update_database_periodically filter

rodrigoprimo committed:
Make context for admin notice documentation more clear

rodrigoprimo commented on includes/integrations/maxmind-geolocation/views/html-admin-options.php:
```diff
@@ -0,0 +1,25 @@
+		<td class="forminp">
+			<fieldset>
+				<legend class="screen-reader-text"><span><?php esc_html_e( 'Database File Path', 'woocommerce' ); ?></span></legend>
+				<input class="input-text regular-input" type="text" value="<?php echo esc_attr( $this->database_service->get_database_path() ); ?>" readonly>
```
In my test, we are showing this path even if the file doesn't exist (before the user enters the key). Should we hide this field when there is no key or if the file doesn't exist? Or maybe change the description below that implies that the file was already downloaded?

ObliviousHarmony replied:
I've added some clarity to the description.

claudiosanches replied:
The idea is to show where the database should be located, this replaces the system status note about it.
Also allows users to download and insert the database in the correct place if they like.

rodrigoprimo reviewed (commented):
I started reviewing this PR, but I wasn't able to finish it today. I'm happy to continue checking it tomorrow if needed.

So far, I read part of the code changes and left some comments. Most of them are just details or questions. My main concern is the wrong hook name when scheduling calls to `woocommerce_geoip_updater` that is probably an issue, but I didn't test to confirm (see inline comment below with more details).

I still need to finish reading the code changes and test this PR.

ObliviousHarmony committed:
Removed the erroneous space on the woocommerce_geoip_updater scheduled event

ObliviousHarmony commented on includes/class-wc-install.php:
```diff
@@ -449,11 +457,8 @@ private static function create_cron_jobs() {
 		wp_schedule_event( time() + ( 3 * HOUR_IN_SECONDS ), 'daily', 'woocommerce_cleanup_logs' );
 		wp_schedule_event( time() + ( 6 * HOUR_IN_SECONDS ), 'twicedaily', 'woocommerce_cleanup_sessions' );
-		wp_schedule_event( strtotime( 'first tuesday of next month' ), 'monthly', 'woocommerce_geoip_updater' );
+		wp_schedule_event( time() + MINUTE_IN_SECONDS, 'fifteendays', ' woocommerce_geoip_updater' );
```
Great catch, I've removed it here and in the update method.

ObliviousHarmony committed:
Added clarity to the file path in the integration settings

claudiosanches commented on includes/admin/class-wc-admin-notices.php:
```diff
@@ -428,6 +430,43 @@ public static function wp_php_min_requirements_notice() {
+	 * @since 3.9.0
+	 */
+	public static function maxmind_missing_license_key_notice() {
+		if ( ! apply_filters( 'woocommerce_maxmind_geolocation_display_notices', true ) ) {
```
Didn't want to hit `get_user_meta()` if it's already disabled by the filter.

rodrigoprimo replied:
Yes, that makes sense as well and it is something that I haven't considered. An alternative that is a bit less legible but takes into consideration the performance concern is:

```php
$filter_dismissed_notice = ! apply_filters( 'woocommerce_maxmind_geolocation_display_notices', true );

if ( $filter_dismissed_notice || get_user_meta( get_current_user_id(), 'dismissed_maxmind_license_key_notice', true ) ) {
	self::remove_notice( 'maxmind_license_key' );
	return;
}
```

claudiosanches commented on includes/admin/views/html-admin-page-status-report.php:
```diff
@@ -435,25 +435,6 @@
 			</td>
 		</tr>
 
-		<?php if ( $settings['geolocation_enabled'] ) { ?>
```
We are removing because it's a 3rd party integration, we it should be moved into a feature plugin in the future.

ObliviousHarmony replied:
A big part of this refactor was a consideration that we move this feature into a plugin. As such, there's no need for it to be in the SSR, as that's a part of core. We've added the path to the integration's setting page to fill this role for support inquiries. Perhaps we might want to add some way for plugins to add to the SSR?

ObliviousHarmony committed:
Added an update task to increase the database version

ObliviousHarmony commented on includes/wc-update-functions.php:
```diff
@@ -2046,3 +2046,20 @@ function wc_update_370_mro_std_currency() {
+ * The TOS for MaxMind's databases require that they be deleted. Since we're making significant changes to
+ * the location and function of the database downloads, let's remove the legacy database altogether.
+ */
+function wc_update_390_remove_maxmind_database() {
```
I agree, added!

ObliviousHarmony committed:
Added a deprecation to the old Geolite library wrapper class' public methods

ObliviousHarmony commented on includes/class-wc-geolite-integration.php:
```diff
@@ -8,6 +8,7 @@
  *
  * @package WooCommerce\Classes
  * @since   3.4.0
+ * @deprecated 3.9.0
```
Good call :) I've gone ahead and added this.

ObliviousHarmony commented on includes/admin/views/html-notice-maxmind-license-key.php:
```diff
@@ -0,0 +1,30 @@
+		echo wp_kses_post(
+			sprintf(
+				/* translators: %s: integration page */
+				__( 'You must enter a valid license key on the <a href="%s">MaxMind integration settings page</a> in order to use the geolocation service.', 'woocommerce' ),
```
This was something @claudiosanches raised as well. The geolocation feature as a whole isn't very well explained, and a lot of people seem to use it when they don't really need to. I'm not sure the best way to go about explaining this to the user though. If we're moving the integration away from core, then the notice would be fine (since it's opt-in to the plugin), but as-is, I can agree we could add clarity to that.

ObliviousHarmony replied:
Notably, different states in the USA have different tax rates though, so I'm not sure how best to explain the use cases where geolocation is a good fit.

rodrigoprimo replied:
Maybe we add something like "Note: if you only sell to one country, you might not need MaxMind geolocation service and, instead of getting a license key, you can disable geolocation in the general settings tab. For more information, see the documentation (link to a page on docs.woocommercecom explaining all the things related to MaxMind geolocation)."

What do you think?

My understanding is that the geolocation provided by WooCommerce core works only to determine the country, so it is not enough for USA stores that need to calculate different taxes depending on the state. But maybe I'm missing something?

rodrigoprimo replied:
@ObliviousHarmony ignore my previous comment. After writing it, I saw your commit where you added some clarity to the message. Looks good to me now. Thanks!

ObliviousHarmony committed:
Added clarification that the MaxMind license key is not strictly necessary

claudiosanches committed:
Improved maxmind_missing_license_key_notice logic

ObliviousHarmony committed:
Changed the MaxMind database update process to move the existing archive instead of deleting it

ObliviousHarmony committed:
Pushed the first woocommerce_geoip_updater execution to fifteen days in the future

This will prevent the routine from deleting the pre-update archive that we just moved in the previous update routine.

claudiosanches committed:
Fixed coding standards
